### ISSUE TITLE
[UX Reliability] Improve the bottom modal animations quality and reliability on native and mWeb platforms

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#49354]
### Problem

Coming from this [thread](https://expensify.slack.com/archives/C05LX9D6E07/p1726571700457229) to bring the NewDot on par with other chat platforms, we need to make sure the modal experience feels fast and the animations are smooth without any lags. Right now the most noticeable lag is happening on the bottom drawers that are often used:
- composer action modal to add attachments or to submit expenses in a chat
- opening or closing the emoji keyboard
- Opening and closing the keyboard - this is not that bad but it also could be improved
- Opening FAB on the main page - the animation is relatively smooth here but its something to monitor too
- Opening more options modal on a comment with long press and then going to edit a message. Saving the edit later. In this flow the modal is opening and closing multiple times with keyboard popping up and down in not smooth fashion

### Solution

Let's ask the @Expensify/design team if we want to explore some non-linear animation easing for these modals. Later (I would say not in the scope of this project), we could explore the bounce effect too for some of these animations.

1. Pick a medium range ios and android device on which we would test
2. Analyse the existing animations, measure its fps and set the baseline of measurements of where we at
3. Research options on what will be the best way to improve this and get us to smooth 60fps animations all the time on the device picked in 1.
4. Discuss the solutions to pick the best one that will work on native and mWeb as well (in the narrow mode where we use the same modal)
5. Implement it and make sure we are able to keep the good animation performance in real world use.

cc @BartoszGrajdek 

<details><summary>Issue Owner</summary>Current Issue Owner: @NicMendonca</details>

**Comment by BartoszGrajdek** [comment]
Hi! I'm Bartosz from SWM and I'll be working on this issue! 😄

**Comment by shawnborton** [comment]
> Let's ask the @Expensify/design team if we want to explore some non-linear animation easing for these modals.

Big fan of that! I think the answer is yes, we do want to explore that.

**Comment by mountiny** [comment]
Added another bullet point to look into, nothing super new, just one more flow to test:
- Opening more options modal on a comment with long press and then going to edit a message. Saving the edit later. In this flow the modal is opening and closing multiple times with keyboard popping up and down in not smooth fashio

**Comment by mountiny** [comment]
@BartoszGrajdek Any updates here 🚀

**Comment by BartoszGrajdek** [comment]
Hey there! 🌟 Apologies for the delayed response; I had a few other tasks pop up, but I dived into the issue yesterday and have a more detailed update for you below. 🙏🏻 

### Devices Update 
I selected iPhone SE 2022 and Oppo A16 initially, but realized the Oppo isn't as mid-range as I thought. I'll swap it out for a better Android device tomorrow at the office.

### Focus on Physical Devices
It's crucial to conduct the final testing and comparisons on **physical** devices. As we well know simulators can sometimes access more resources than actual phones do, skewing the results. 📱

### Baseline of measurements
I have recorded the animations on **mWeb** and **Android**, but I'm still struggling with getting the **iOS**  to build on a physical device, that's something I'll focus on today. 😅 

Tomorrow once I swap the Android phone with a better one I'll try to summarise our current state of these animations in terms of fps - given I get the iOS to build 🍎 

### Next Steps - Action Plan
1. Check which parts of the code are responsible for these specific animations
2. Make sure that all animations are using Reanimated and not react-native's Animated 
3. Analyse if any obvious bugs could influence the performance of animations
4. If issues persist, explore different ways to improve performance and discuss options with the team.

We'll also likely get another person from the team to help me out here to speed this process up 🚀 

Keep me posted on any thoughts or recommendations, and let's keep the momentum going! If you have questions or need more details, feel free to reach out. I'll make sure to keep you updated regularly 🙌🏻 

Looking forward to our progress! 💪🏻

**Comment by mountiny** [comment]
Lovely, thanks for the focus and update

**Comment by RachCHopkins** [comment]
Hey @mountiny this may be a silly question, but do you actually need me on this issue for anything?

**Comment by mountiny** [comment]
Nah all good

**Comment by BartoszGrajdek** [comment]
Hello again! 🌟 I'm back with a detailed update on my activities over the past few days. 🙌🏻 

### Recordings & Baseline Measurements

I’ve completed the remaining recordings, including the native iOS recordings on an iPhone 11. After discussing with the Reanimated team, we've concluded that there isn't an *ideal* way to measure the FPS of animations. For now, I'll compare changes using the number of unique frames in the recorded videos. 

In these results, I’ve excluded frames where only the background opacity changes. I believe our primary focus should be on improving the **smoothness** of the moving modal, as it currently performs the worst. 😓

| | iPhone SE mWeb | iPhone 11 Native | Realme GT2 mWeb | Realme GT2 Native |
| -- | -- | -- | -- | -- |
| Attachments Modal | 6 | 4 | 6 | 4 |
| Emoji Picker | 4 | 3 | 4 | 3 |
| FAB | 15 | 10 | 15 | 13 |
| Edit Modal | 11 | 9 | 12 | 15 |

*These results indicate the number of unique frames in each animation.* 📊

### Research Results

After recording and comparing all the results, I've started investigating the cause of the poor animation performance. I found out that all these components use the `BaseModal` component, which in turn utilizes the `react-native-modal` library. 

I suspect that the `react-native-modal` library is the **culprit** behind the significant lag. I believe that the issue is the fact that `react-native-modal` relies on `react-native-animatable` instead of `react-native-reanimated`. `react-native-animatable` is essentially a wrapper for RN Animated, which could be slowing down the animations considerably. 🐢

### Next Steps

To **verify** my suspicion, I plan to locally replace `react-native-animatable` in `react-native-modal` with Reanimated. I will then compare both versions to determine if this is indeed causing the issue. 👀 

### Possible Outcomes

If my proposed solution does not resolve the issue, I will delve deeper to identify other potential causes.

If the solution is effective, there are a few paths we could take:
1. Apply my solution as a patch. 🩹
2. Create a fork of `react-native-animatable` that uses Reanimated (the library maintainers are hesitant to change the animation lib)
3. **Take over** the library, which hasn’t been updated in two years. 
4. Replace the library with another modal library that uses Reanimated. 
5. **Develop our own Modal component** (I think this may be an overkill)

### Your Input

If you have any suggestions or feedback on this matter, *please let me know*. 

I'll make sure to give you an update once I have more definitive results or progress to share. Stay tuned! 🔜

**Comment by BartoszGrajdek** [comment]
I'll add an update here later today 😄

**Comment by BartoszGrajdek** [comment]
Hi! 🌟 

I've put together a PoC using `react-native-reanimated` in `react-native-modal` instead of `react-native-animatable` and also took a closer look at each individual animation. 

I've got a rough outline of the possible next steps we can take to improve all the animations and I'll put together a detailed summary with all this information to share it and discuss with you tomorrow. This summary will include potential strategies and priorities for moving forward. 🚀

The good news is that @zfurtak from our SWM team will be helping me with these improvements, which should help us speed up this process. 

Looking forward to making these enhancements! ✨

**Comment by mountiny** [comment]
Amazing, what a team, thanks for an update

**Comment by BartoszGrajdek** [comment]
Hi! ✨

Over the past few days, I've built a PoC using `react-native-reanimated` instead of `react-native-animatable` in the modal library. I've also thoroughly examined each animation on our list. There's a lot to discuss, so let's start with changes we can implement in the app now, without swapping the library. 👀

### Transition Times
A key issue with our current animation setup is how transition times are defined in certain cases. Let's explore them:

**Attachment Modal:** `animationInTiming: 100ms` & `animationOutTiming: 300ms` – I think that it was initially set for an attachment modal with just 2 menu items and this timing no longer fits as different parts of ExpensifyApp now have varying item counts. I suggest adjusting the transition time to one relative to the screen and modal size ratio, ensuring a smoother experience across all different kinds of modals.

**Edit Modal:** `animationInTiming: 300ms` & `animationOutTiming: 1ms` – This disabled out animation could be linked to the keyboard opening when selecting “_Edit comment_.” My idea is to enable animations when the user isn't actively editing a comment.

**Emoji Modal:** `animationInTiming: 1ms` & `animationOutTiming: 1ms` – The lack of smooth animation on iOS led to its disabling. With the library replaced, it's working much better, still a bit jittery in the PoC—but this is fixable!

### Keyboard Behaviour
@zfurtak has taken on this task. We've analysed various keyboard interactions within the app and found several improvements we can implement for a better experience. Expect an update from Zuzanna soon once there's progress to share. 🔜 

### General Animations Flow
The current flow for modals begins with the bottom modal moving upwards, followed by the backdrop's opacity change after ~100ms. This results in an uneven animation, which affects the UX. 🐢

I'm sure there are more examples of changes we could make in the flow of animation itself, which we'll discuss and address individually (especially keyboard animations). These enhancements could significantly improve the user experience with (hopefully) minimal changes.

### Frame Measurements
During the PoC development, I realised that the recordings from physical devices aren’t reliable for comparing the animation performance. FPS can fluctuate—for example, an iOS screen recording may start with 40 frames in the first second and drop to 23 by the 30th second—making accurate comparisons difficult.

We'll still use physical devices for testing but not as an exact gauge of animation quality. Because of this, I've decided to take new measurements using the iOS simulator and Android emulator (captured with macOS's built-in screen recordings, consistently at 60fps).

### Reanimated PoC
As mentioned, I've created a PoC of `react-native-modal` with `react-native-reanimated`. The animations are hard-coded and need further work, but the results are promising. It's possible to replace the animation library entirely, sparing us the hassle of switching libraries or dealing with potential regressions.

The aim was to determine if `react-native-animatable` in the modal library caused low FPS, which I believe we proved. There's much more to discuss on this topic, so I'll create a separate discussion thread on **#newdot-performance** for broader community input.

I compared results from both implementations using the method described above—note that improvements are still possible and the solution is not complete. I’ve marked the results that are using reanimated with (R), the values represent the number of frames the animation had. I didn't compare mWeb since it only involves opacity changes, where FPS differences have less impact than slide-in animations.

Animation (inTiming, outTiming) | iOS Native | iOS Native (R) | Android Native | Android Native (R) | Ideal frames
-- | -- | -- | -- | -- | --
FAB (300/300) | 12 / 5 | 18 / 18 | 13 / 6 | 15 / 15 | 18 / 18
Attachments Modal (100/300) | 4 / 4 | X | 5 / 5 | X | 6 / 18
Attachments Modal (300/300) | 11 / 5 | 13 / 15 | 13 / 5 | 16 / 15 | 18 / 18
Emoji Picker (1/1) | 1 / 1 | X | 1 / 1 | X | 1 / 1
Emoji Picker (300/300) | 13 / 5 | 14 / 17 | 13 / 6 | 16 / 16 | 18 / 18
Edit Modal (300/1) | 12 / 1 | X | 13 / 1 | X | 18 / 1
Edit Modal (300/300) | 11 / 7 | 18 / 17 | 13 / 7 | 17 / 16 | 18 / 18


> [!NOTE]
> Since the backdrop opacity and drawer slide-in animations don't always sync up, I'm only counting the frames when the slide-in animation is happening (animatable). Ideally, these would align, but I don’t want unsynchronised timing to skew the results.

### **Next Steps**

Here’s what we’ll focus on moving forward:

- Find all the other places where bottom modals could be used to ensure that the quality of our solution is the same across the entire app (@zfurtak)
- Discuss how we’re going to proceed with the implementation using `react-native-reanimated` as a library responsible for bottom modal animations (@BartoszGrajdek)
- Explore how to adjust transition times for various animations and provide examples to the design team (@BartoszGrajdek)
- Investigate the current keyboard behavior within the app, compare it to other apps, and propose solutions to enhance user experience (@zfurtak)

We'll likely address the second, third, and fourth tasks in separate PRs, given that they are distinct topics.

That's it for today! I hope this sheds some light on our progress. Let me know if you have further suggestions or questions. I'm excited to keep pushing this forward!

More updates coming soon! 🚀

**Comment by BartoszGrajdek** [comment]
Added a message to #newdot-performance here: https://swmansion.slack.com/archives/C05LX9D6E07/p1729074576837529

**Comment by mountiny** [comment]
Discussing in slack

**Comment by zfurtak** [comment]
Hello 🙋‍♀️ Time for my...
### ✨ Analysis of keyboard behaviour ✨

After analysing the cases in the app where we use the keyboard, I have come to a few conclusions. Compared to different apps like Messenger, WhatsApp or Slack, the keyboard behaviour in Expensify is not as consistent and there is ample space for improvement. 📈

There are situations where the keyboard behaves poorly, such as appearing without sliding, just jumping up and down. These issues simply need fixing. Here is one example:

https://github.com/user-attachments/assets/7a8699d8-9cdd-483f-a780-51546f75c3b8


### Mechanisms for improving user experience ⌨️

I have two suggestions to enhance the user experience with the keyboard: 

There are cases where the keyboard opens and affects components at the bottom of the screen (e.g., a `Button`), if you watch carefully you’ll see that the button jumps from the bottom of the screen to the middle and "waits" for the keyboard to arrive. My suggestion is to make the button (and in general all components like this) move smoothly with the keyboard. Good example for this can be keyboard behaviour from Slack (attached below)

https://github.com/user-attachments/assets/7e09d85b-805d-40a3-91f2-edcd6c454af1

https://github.com/user-attachments/assets/49c1cccf-78e7-450c-a7ac-f8797a1bf1e3
 
---
Another observation is the compatibility of the chat page with the keyboard. There are situations where the chat moves up and down based on the keyboard's position, which is not visually appealing. Like in the example below:

https://github.com/user-attachments/assets/48dab59a-9b37-4c5d-b140-bfcb338d509d

For this case I made a comparison with Messenger which handles the situation really nice, see below 🔍

https://github.com/user-attachments/assets/77da3e5e-8e5a-47e6-8db9-9a08a7b2e194

https://github.com/user-attachments/assets/33118fb3-ca09-40bc-9291-a9af5f20c791

For a better user experience, I’d recommend keeping the chat fixed in place, allowing the keyboard to appear seamlessly, as it is only a brief moment and the background is blurred while the modal is open.

Let me know what you think! 😊

**Comment by dannymcclain** [comment]
Wonderful thoughts and examples laid out there. Thanks for the thoughtful post!

**Comment by dubielzyk-expensify** [comment]
Lovely set of thoughts and agree with all of them. I believe all the apps you're pointing to are native apps (?). Are all these fixes outlined doable with React Native?

**Comment by zfurtak** [comment]
I'm not entirely sure as I've never really looked into keyboard behaviour so precisely, but I'll dive into the code and see what can be done. Some of the bug fixes seem like they should be straightforward.
 I'll keep you posted on my progress 😊

**Comment by mountiny** [comment]
Same i think we all agree we want to align with these industry leading apps. 

I believe @kirillzyusko would be great resource for anything keyboard related. Do  you think the aforementioned improvements are achievable in react native?

**Comment by mountiny** [comment]
What is holding https://github.com/Expensify/App/pull/42143? it seems like there are no updates for a month

**Comment by zfurtak** [comment]
> It has been fixed (but not merged yet) in: https://github.com/Expensify/App/pull/42143

Great! Then I will wait until it's merged and then maybe introduce some other improvements 💪

> Before (on paper architecture) we had smooth animations. Then on Fabric something stopped work. @zfurtak do you test debug or release variant? I remember someone told me that KeyboardAvoidingView is broken in debug mode, but works fine in release (I think it was truth in the past, but I haven't verified it with latest RN version).

I tested it in debug mode on simulator as well as in the app downloaded from App Store, the `Button` behaves similarly (it doesn't slide, just jumps into right place). So I guess the problem is relevant in all versions.

> @zfurtak not sure what you meant here... We don't blur a background at the moment (we just show an overlay with semi-transparent black color). Do you wnat to change a design and blur background when modal is shown?

No no don't worry... Semi-transparent black background was exactly what I meant, just used wrong words 😅

> We also suggested such option (context menus instead of bottom sheet) a really long time ago (a year ago or even more), but UX/UI team didn't want to change the current behaviour (from what I remember).

Maybe we can rethink it in this issue 🤓

**Comment by kirillzyusko** [comment]
> What is holding https://github.com/Expensify/App/pull/42143? it seems like there are no updates for a month

@mountiny I think Taras has been working on it, but then he got more important tasks and switched to them. I'm planning to work on fixing e2e tests this week (because it's quite critical stuff) and if Taras is still busy, then I'll take over this PR and will try to bring it to finish line 💪 So I hope it should be merged soon 😊

**Comment by mountiny** [comment]
We still have lots to do so removing the ready for payment from here

**Comment by zfurtak** [comment]
hi @kirillzyusko how is the work on the PR going? 😊

**Comment by kirillzyusko** [comment]
@zfurtak I slightly overestimated my capabilities and e2e were broken in much more places than I initially expected 😔 So this week I was working mostly on e2e stuff, but I am planning to switch to https://github.com/Expensify/App/pull/42143 early next week 👀

**Comment by BartoszGrajdek** [comment]
Hi! 👋🏻  

Here's a short update on `react-native-modal` changes 👀  

I've refactored the source code and it now follows most of our style guidelines. That included a transition from class component to a functional one, which together with the removal of `react-native-animatable` has caused some regressions - luckily all of these problems should be solved now and we can focus solely on reanimated performance. ⏩ 

### Next steps
@sumo-slonik is currently helping me with some code clean-up and Typescript, while I'm focusing on improving animations & handling all the modal use cases across E/App. We're hoping to create a PR next week, but I'll keep you updated on that. 🚀 

More updates coming soon! ✨

**Comment by mountiny** [comment]
Thanks

**Comment by BartoszGrajdek** [comment]
Hello! 🙌🏻 

Posting a short update on the progress of the `react-native-modal` transition, a longer analysis coming later this week/next week. 🤞🏻  The situation is changing dynamically, @sumo-slonik is now working on a different issue, and I haven't been working full-time since the last message, that's why it slowed down a little, but now I'm fully available and Kuba is wrapping up the other issue which should help speed things up. ⏩ 

Our plan is to first address all bottom-docked modals (in the first PR), since that was the main goal of this issue and then update all the other modal cases, that way we'll methodically switch from `react-native-modal` to our solution while avoiding any unwanted regressions/reverts. 🙏🏻 

We experienced some bottlenecks during the development in the last 2 weeks. After some trial and error as well as discussions with our open-source team, they pointed out that the problems we experienced with the reanimated animations are likely related to performance issues. They seem to come from `PopoverMenu` (long rendering time) and we're currently working on finding a solution that will allow us to continue. 

The code we refactored is in a good state, there's one bug with `SlideInDown` we found, but the performance limitations are the main reason we weren't able to push the first PR yet since they affect the results.

I'm hoping we can share a first PR with you before the fix for `PopoverMenu` is ready to progressively improve the modal animations, but that's something I'll update you on once we manage to find a solution to the `SlideInDown` bug.

Let me know if you have any questions, more coming soon! 🚀

**Comment by BartoszGrajdek** [comment]
Hello again! ✨

I'm back with another update, and I'm happy to report that we’re moving forward! 🚀 I believe we’ll have the PR ready for review in the week when SWM returns from being OOO.

A quick note on our current team capacity: Kuba is occupied with some other tasks, which means I'm working solo on this issue currently. Additionally, I'm wrapping up a task in **`react-native-live-markdown`**, so it's a busy time 😅 That means that the performance-related issues and the analysis I had mentioned will have to wait for now, as my focus is on getting this first PR through to production. 

Over the past two weeks, I've resolved the **`SlideInDown`** bug that we encountered, as well as improved the modal animations to make up for the `PopoverMenu` performance problems. Currently, I'm thoroughly testing every bottom-docked modal in our app to ensure everything is functioning correctly 👀

Once I've verified all use cases and fixed any arising issues, I'll create a PR for the C+ team to test. Fingers crossed, all goes smoothly! 🤞🏻

I'll keep you posted if I encounter any significant issues. Excited to bring this to a close soon! 🥳

*PS. SWM will be OOO from December 23 to January 5. However, I may be working on some of these days*

**Comment by mountiny** [comment]
Lovely, thank you!

**Comment by dannymcclain** [comment]
>I also have a few questions to the design team about these animations, not sure where can I post it - is this issue fine, or should I start a thread somewhere on Slack? 🤔

This issue is probably fine, but starting a thread somewhere in Slack might be better for faster collaboration. Just make sure to tag `@designteam` in Slack. cc @Expensify/design

**Comment by dubielzyk-expensify** [comment]
Excited to see more 😄  Happy to answer any questions like Danny said

**Comment by BartoszGrajdek** [comment]
Created a 🧵 in newdot-performance 😄 

https://swmansion.slack.com/archives/C05LX9D6E07/p1736848147574949

**Comment by BartoszGrajdek** [comment]
Hello everyone! cc @mountiny 

Sorry for the delay, but with the search team using single-line live markdown for the first time in E/App there are just too many issues at hand 😅 

Anyway, I’ve talked with our reanimated performance expert and I’d like to discuss our next steps here! 🚀 

### Current State
I have a draft PR that will add the refactored code into all _popover-related bottom modals_, I’ve tested it quite well, so there shouldn’t be any significant issues with it. I do need to address the review comments from my colleagues and some small bugs but overall it should be good to go. The main blocker currently is that we’re still discussing how we want the animations to look with the design team, but I’m hoping that we’ll get it solved in the coming days. 🤞🏻 

### Performance
The biggest hurdle over the span of this entire project was that there was no perfect way to measure the FPS of animations deterministically. You can’t record it on a physical device, since it sometimes optimizes the screen capture (so it’s often not 60fps), or gets overwhelmed by it if you actually do get a 60fps recording. I’ve also tried measuring the FPS by recording the simulator with built-in macOS screen capture, but that’s also not ideal - from my experience the simulator works worse compared to the physical device in terms of animation performance (even if the UI thread stays at 60fps), and trust me running a release build every time I’ve made a change wasn’t the most pleasant… 😖  For these reasons I’m not going to focus on the exact measurements of FPS that much. With `react-native-reanimated` in place we should be running entirely on UI thread which is consistently at 57-60FPS during the animations, it’s also likely that on higher-end devices with 120Hz refresh rate, we could even achieve >60FPS since we’re no longer limited by the technical restraints here ⚡ 

What I can tell you is this: after all the improvements, the animations feel better. I was testing it using lower-end of mid-tier devices on a release (**android**) and debug (**iOS**). After showing it to other people at SWM they all said that the animations feel smooth, which is what we’re aiming for :dart: I've recorded some before/after demo videos on physical devices for you, which I'll attach at the end (iOS is in debug, so it should be even better in release). Keep in mind that the animation configuration will probably change once we decide how we want the animations to look. Let me know what you think! 👀 

### Limitations
The main limitation is the `EmojiPicker` which is a heavy component and it’s difficult to animate it smoothly. There are just too many SVGs/images rendered and moved at once, which is not the easiest task. This isn't a big deal on iOS, since it has a better performance compared to Android in general so the animations are smooth, but it certainly could be better on Android. There are some tricks we could try (i.e. Facebook Messenger only loads the first row of emojis, and only in the last ~2 frames loads the rest), I’ll see what we can do with what we have for now if that doesn't work out we can create a separate issue to work on this specifically 😁 

### Next Steps
We’re planning to replace all bottom-docked modals and there are quite a few of them. There are 6 main generic components: `ConfirmModal` (~70) `DecisionModal` (9) `SelectionListWithModal` (12) `Popover` (5) `PopoverWithMeasuredContent` (3) `PopoverMenu` (9). I’ve already tested most of them and there shouldn’t be any problems, but for the sake of our sanity in case there are any regressions/changes we’ll do it in a few batches, the first being Popover-related as I’ve mentioned and that’s intentional, because all modals mentioned in the issue description are in this batch. Also, Błażej said he’ll be able to help me out with the next few PRs to make the work easier 🤝 

There's also a problem with easing that I found, which is that it is only applied to layout animations once and when you reopen a modal it just uses linear easing. I'll try talking to the reanimated team about it, it's just something I found out yesterday and I didn't have a chance to show it to anybody. If it's something more complicated to fix we'll just add these improvements with linear easing & later include the fix.

From my discussion with the reanimated expert, he did notice some things we could improve, but as he said they aren't a big deal, and if the performance won’t be a problem anymore they're probably not worth investigating. It is something we could look into in the future though…

My question to you is regarding gestures - the old library was using built-in `react-native` gestures and we did refactor that code, at some point I was even making sure it works everywhere. The thing is considering we’re moving the code to E/App we should probably migrate it to `react-native-gesture-handler` to keep things consistent. The problem is that although at first, we were planning to handle all modals using our new code, the scope has changed with Chris’s issue and we’re now only going to use it in bottom-docked modals, which AFAIK don’t use gestures anywhere. Personally, I’d just delete the code, but that’s not my call. WDYT I should do? 🤔 @mountiny 

<details>
<summary>Comparison Videos</summary>

### iOS

<table>
<tr>
 <td>Before
 <td>After
<tr>
 <td>

https://github.com/user-attachments/assets/2b555d40-f3a8-4b9c-85d8-8a18b4024f53


 <td>

https://github.com/user-attachments/assets/607d099f-a7b6-4935-a966-b227e2483e8f


</table>


### Android

<table>
<tr>
 <td>Before
 <td>After
<tr>
 <td>

https://github.com/user-attachments/assets/96366105-d2be-40ee-9333-9c4e89610212


 <td>

https://github.com/user-attachments/assets/2c729495-8281-4e1d-aa07-06c9cee524b3


</table>


</details>

**Comment by mountiny** [comment]
Thanks for the write up.

Lets please start with the Popover modals one by one, only one type per PR so we can test and revert easily if needed

> My question to you is regarding gestures - the old library was using built-in react-native gestures and we did refactor that code, at some point I was even making sure it works everywhere. The thing is considering we’re moving the code to E/App we should probably migrate it to react-native-gesture-handler to keep things consistent. The problem is that although at first, we were planning to handle all modals using our new code, the scope has changed with Chris’s issue and we’re now only going to use it in bottom-docked modals, which AFAIK don’t use gestures anywhere. Personally, I’d just delete the code, but that’s not my call. WDYT I should do? 🤔

Yeah, if its not needed for now lets remove it and we can add the gesture handling code back in whenever needed

**Comment by mountiny** [comment]
In terms of the easing curve, I will leave that for @Expensify/design from the videos the After felt a tad bit too slow for me, but also we will see how it feels in the real app

**Comment by BartoszGrajdek** [comment]
> In terms of the easing curve, I will leave that for @Expensify/design from the videos the After felt a tad bit too slow for me, but also we will see how it feels in the real app

Yeah, I agree, but in all of the after videos the curve used was _linear_ which is probably why it has a different feel. We'll decide on our ideal solution in the thread mentioned above.

**Comment by BartoszGrajdek** [comment]
I've created 3 draft PRs for the remaining cases described in this issue's description. 

After #55157 gets merged we can follow up with the remaining modals. Once that's completed we'll migrate other modals in batches. 

That will allow us to gradually switch to the new solution and it's what @mountiny suggested 🙌🏻

**Comment by dannymcclain** [comment]
Personally I do think that's worth fixing. Every little unexpected jump in the UI has a negative impact on the user experience. `:my-2-cents:`

**Comment by mountiny** [comment]
Yeah but also we still work on some changes to the safe padding and keyboard handler with Margelo and @kirillzyusko @chrispader maybe their changes will help with this issue too

**Comment by chrispader** [comment]
> Yeah but also we still work on some changes to the safe padding and keyboard handler with Margelo and @kirillzyusko @chrispader maybe their changes will help with this issue too

Working on finishing up the PR(s) for the bottom safe area + padding handling right now. Might take some time to merge and QA them though, since we have to merge two PRs consecutively, in order to not break the bottom safe area padding on screens.

> When the keyboard is being opened/closed, we're removing/adding the SafeAreaBottomPadding, making it feel like the modal height is jumping.

Our PRs will not fix this issue though, since optimally we would need to create an animation of the bottom safe area padding while the keyboard is opening. The current logic is valid, but it's just not very aesthetic, since the padding is only removed after the keyboard opened fully and vice versa with closing the keyboard.

Maybe @kirillzyusko can shed some more light on how/if we can make this smoother with `react-native-keyboard-controller`.

**Comment by chrispader** [comment]
Also just leaving a note about [this PR](https://github.com/Expensify/App/pull/56219) which migrates all `AttachmentModal`s to screens. We'll definitely want to migrate those modals, since they always behave like screens (right-to-left) and never modally.

Designing `AttachmentModal`s as modals was originally just for DX convenience, since we share some code in the wrapping screen/component, that couldn't be shared easily with other screens through navigation state/params. My PR solves this completely

**Comment by kirillzyusko** [comment]
> Maybe @kirillzyusko can shed some more light on how/if we can make this smoother with react-native-keyboard-controller.

I've fixed this problem in https://github.com/Expensify/App/pull/54533 but this PR got reverted (because introduced some weird problems - you weren't able to select any emoji 🤷‍♂ ).

So yeah, we can make it smoother with `react-native-keyboard-controller` - just need to figure out why applying `translateY` makes `Touchables` unresponsive 🤔 

Also I know that @danstepanov was going to work on this problem too, so maybe he can shed more light 👀

**Comment by shawnborton** [comment]
Totally agree with Danny here, let's definitely fix that/smoothen it out.

**Comment by dubielzyk-expensify** [comment]
Agree especially given the goal is to improve the animation quality in the first place.

**Comment by mountiny** [comment]
@BartoszGrajdek @zfurtak Great job getting the first PR merged!

**Comment by BartoszGrajdek** [comment]
Thanks! I'm currently looking into the FAB issue to enable the new animations there once again 😄

**Comment by BartoszGrajdek** [comment]
Added a PR fixing the FAB issue [here](https://github.com/Expensify/App/pull/57536) I've tested it in other modals as well and didn't find any regressions.

**Comment by mountiny** [comment]
@BartoszGrajdek @zfurtak @blazejkustra What are the next steps on this issue? Thanks

**Comment by BartoszGrajdek** [comment]
We were waiting for the FAB issue to pass QA tests first before pushing new PRs, but that's done now. We have 3 draft PRs ready for all of the other use cases mentioned in this issue description and after that, I believe we should be prepared to go through each modal type one by one.

FYI @blazejkustra will be working on this issue now

**Comment by mountiny** [comment]
@blazejkustra i know you busy with the live helpdot project, anyone else able to to help here?

**Comment by blazejkustra** [comment]
Not really, everyone is busy with their current tasks 😮‍💨 I'll check with the team on Monday.

**Comment by ZhenjaHorbach** [comment]
At the moment, payment should be hold
Since this issue includes several PRs that we are working on !

**Comment by mountiny** [comment]
Waiting for someone else from SWM to step in

**Comment by blazejkustra** [comment]
> Waiting for someone else from SWM to step in

I'm afraid everyone is busy and won't be able to pick this up for the next two days 😢

I'll post an update on Wednesday if anyone becomes available.

**Comment by blazejkustra** [comment]
> I'll post an update on Wednesday if anyone becomes available.

Sorry for not updating sooner, but still - no one is available. I'd love to get back to this ASAP, but Side Pane bugs keep piling up. 😢

**Comment by blazejkustra** [comment]
I've almost finished fixing all the side pane bugs and should be able to get to this later this week.

**Comment by blazejkustra** [comment]
I just wrapped up all the Side Pane bugs, so tomorrow is the day to revisit this! 🙀

**Comment by mountiny** [comment]
How exciting!

**Comment by ZhenjaHorbach** [comment]
Not overdue

**Comment by blazejkustra** [comment]
Update: I'm working on EmojiPicker PR

**Comment by blazejkustra** [comment]
@kirillzyusko Any updates on [this](https://github.com/Expensify/App/issues/49354#issuecomment-2663328560)?

The jumping issue is quite noticeable both on main and after my changes. You can see it happening on native platforms [here](https://github.com/Expensify/App/pull/59539#issuecomment-2772709164).

**Comment by kirillzyusko** [comment]
> Any updates on https://github.com/Expensify/App/issues/49354#issuecomment-2663328560?
>
> The jumping issue is quite noticeable both on main and after my changes. You can see it happening on native platforms https://github.com/Expensify/App/pull/59539#issuecomment-2772709164.

Actually no updates 😔 One thing I would try is to apply translation (translateY) on a different View (i. e. somewhere higher). Or replicate changes and verify it they have the same bug (maybe it was RN problem and recently RN was updated, right)?

**Comment by blazejkustra** [comment]
I'll be OOO from April 11-16. Once I'm back, I'll continue migrating remaining modals 🚀 

Update: I created three new PRs so that @ZhenjaHorbach has something to do while I'm away! I've tested them on all platforms, and I don't anticipate any major issues or regression coming out 😄 

While I'm out, I've asked @borys3kk to keep an eye on these. If the main branch goes out of sync or any other issues arise, please tag him 🙏

[[UX Reliability] Use new modal in BasePopoverReactionList #60016](https://github.com/Expensify/App/pull/60016)
[[UX Reliability] Use new modal in ConfirmModal #60023](https://github.com/Expensify/App/pull/60023) - These are used throughout the app, but their usages are very similar, and migrating each one individually would be quite time consuming. I hope you agree it’s okay to handle all the confirm modals in one PR, @mountiny?
[[UX Reliability] Use new modal in DecisionModal #60025](https://github.com/Expensify/App/pull/60025) - Same as above, wdyt @mountiny?

**Comment by ZhenjaHorbach** [comment]
> I'll be OOO from April 11-16. Once I'm back, I'll continue migrating remaining modals 🚀
> 
> Update: I created three new PRs so that [@ZhenjaHorbach](https://github.com/ZhenjaHorbach) has something to do while I'm away! I've tested them on all platforms, and I don't anticipate any major issues or regression coming out 😄
> 
> While I'm out, I've asked [@borys3kk](https://github.com/borys3kk) to keep an eye on these. If the main branch goes out of sync or any other issues arise, please tag him 🙏
> 
> [[UX Reliability] Use new modal in BasePopoverReactionList #60016](https://github.com/Expensify/App/pull/60016) [[UX Reliability] Use new modal in ConfirmModal #60023](https://github.com/Expensify/App/pull/60023) - These are used throughout the app, but their usages are very similar, and migrating each one individually would be quite time consuming. I hope you agree it’s okay to handle all the confirm modals in one PR, [@mountiny](https://github.com/mountiny)? [[UX Reliability] Use new modal in DecisionModal #60025](https://github.com/Expensify/App/pull/60025) - Same as above, wdyt [@mountiny](https://github.com/mountiny)?

I kept hoping that you would stop with new PRs

<img src="https://github.com/user-attachments/assets/4eb3eccb-0902-41c8-bd84-f4c2ad113405" width="300"/>

Have a nice rest!

**Comment by mountiny** [comment]
They keep coming, the plan sounds good to me, thanks!

**Comment by ZhenjaHorbach** [comment]
At the moment, payment should be hold
Since this issue includes several PRs that we are working on !

**Comment by zfurtak** [comment]
Hi guys 🙋‍♀️
I'm back on this issue,  helping Błażej raise PRs faster  🏎️
One with [`RequireTwoFactorAuthenticationModal`](https://github.com/Expensify/App/pull/60495) is ready for a review.

**Comment by zfurtak** [comment]
[`SearchSelectedNarrow`](https://github.com/Expensify/App/pull/60499) also ready for a review 🤓

**Comment by mountiny** [comment]
@ZhenjaHorbach Kicked off the builds for both PRs

**Comment by ZhenjaHorbach** [comment]
> [@ZhenjaHorbach](https://github.com/ZhenjaHorbach) Kicked off the builds for both PRs

This will be useful
Thanks !

**Comment by jmusial** [comment]
[SelectionListWithModal](https://github.com/Expensify/App/pull/60829) is ready for review

**Comment by jmusial** [comment]
[New modal on SearchList](https://github.com/Expensify/App/pull/60961) is ready for review

**Comment by jmusial** [comment]
New modal in [MoneyRequestReportTransactionList](https://github.com/Expensify/App/pull/60952), [Download PDF on ReportDetailsPage](https://github.com/Expensify/App/pull/60956), [components using Popover](https://github.com/Expensify/App/pull/61165) are ready for review. Those are the last ones :)

**Comment by ZhenjaHorbach** [comment]
> ⚠️ Looks like this issue was linked to a **Deploy Blocker** [here](https://github.com/Expensify/App/issues/62682)
> 
> If you are the assigned CME please investigate whether the linked PR caused a regression and leave a comment with the results.
> 
> If a regression has occurred and you are the assigned CM follow the instructions [here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#regressions).
> 
> If this regression could have been avoided please consider also proposing a recommendation to the PR checklist so that we can avoid it in the future.

I'm not sure if this is our regression
Since the regression appeared after adding an input in FeatureTrainingModal that we updated a month ago
But since we are responsible for the new version of modals
I suppose we should fix this

**Comment by mountiny** [comment]
I think the AttachmentModal might be already handled by @chrispader

**Comment by blazejkustra** [comment]
> I think the AttachmentModal might be already handled by [@chrispader](https://github.com/chrispader)

@chrispader what's the status of `AttachmentModal` migration to a screen?

**Comment by blazejkustra** [comment]
> [@blazejkustra](https://github.com/blazejkustra) Would you be able to get some measurements to see if the performance will be better using reanimated for those modals too? so we can wrap this P/S around improving performance

Yes, we'll do some research with @jmusial 👍

**Comment by ZhenjaHorbach** [comment]
No payment required yet
Since we have one last a PR to review

**Comment by ZhenjaHorbach** [comment]
I think we are done here 😅
Or no 
Following [this thread](https://expensify.slack.com/archives/C05LX9D6E07/p1750147058498549) ? 🧐

**Comment by mountiny** [comment]
That would be a new issue, but I think we can pay it out here. I'm not sure if any new special regression tests are required since this is technically just a performance improvement with no change in expected results

@ZhenjaHorbach Could you please summarize what needs to be paid out and has not been paid out yet? thanks

**Comment by ZhenjaHorbach** [comment]
> That would be a new issue, but I think we can pay it out here. I'm not sure if any new special regression tests are required since this is technically just a performance improvement with no change in expected results

Agree 
A specific feature for regression tests 😅

> [@ZhenjaHorbach](https://github.com/ZhenjaHorbach) Could you please summarize what needs to be paid out and has not been paid out yet? thanks

There have been no any payments yet
But today I will count how much PRs there were in this issue !

**Comment by ZhenjaHorbach** [comment]
Okay 
I found 19 PRs which I reviewed

https://github.com/Expensify/App/pull/55157
https://github.com/Expensify/App/pull/56605
https://github.com/Expensify/App/pull/56606
https://github.com/Expensify/App/pull/59539
https://github.com/Expensify/App/pull/60025
https://github.com/Expensify/App/pull/60016
https://github.com/Expensify/App/pull/60023
https://github.com/Expensify/App/pull/60495
https://github.com/Expensify/App/pull/60499
https://github.com/Expensify/App/pull/60719
https://github.com/Expensify/App/pull/60829
https://github.com/Expensify/App/pull/60952
https://github.com/Expensify/App/pull/60956
https://github.com/Expensify/App/pull/60961
https://github.com/Expensify/App/pull/61165
https://github.com/Expensify/App/pull/61372
https://github.com/Expensify/App/pull/61838
https://github.com/Expensify/App/pull/62397
https://github.com/Expensify/App/pull/62799

And 7 regressions !

https://github.com/Expensify/App/pull/57536
https://github.com/Expensify/App/issues/59615
https://github.com/Expensify/App/issues/59769
https://github.com/Expensify/App/issues/60977
https://github.com/Expensify/App/issues/61951
https://github.com/Expensify/App/issues/61986
https://github.com/Expensify/App/issues/62175

**Comment by mountiny** [comment]
@ZhenjaHorbach Can you please confirm if any of the PRs been to fix the regressions listed?

**Comment by ZhenjaHorbach** [comment]
> [@ZhenjaHorbach](https://github.com/ZhenjaHorbach) Can you please confirm if any of the PRs been to fix the regressions listed?

I did't add regression PRs to the list which I reviewed

**Comment by ZhenjaHorbach** [comment]
But wait 
Maybe only this 
https://github.com/Expensify/App/pull/62397
We made global changes in [this PR](https://github.com/Expensify/App/pull/60952)
But after [this regression](https://github.com/Expensify/App/issues/62175) we decided to revert a part of changes and make updated changes in a separate PR

But otherwise I did't add regression fixes to the list of PRs which I reviewed

**Comment by mountiny** [comment]
Alright, thank you! This was a long issue that lasted many months, so it seems like we have 18PRs—18*250 = 4500 now minus 7 regressions ($125 each), which is 7*125 = 875, so the total is 4500-875=$3625.

The payment total for @ZhenjaHorbach for this entire initiative would be $3625. This seems reasonable, given that the project took 8-9 months to complete.

@NicMendonca, could you help us with the payout, please?

**Comment by ZhenjaHorbach** [comment]
> Alright, thank you! This was a long issue that lasted many months, so it seems like we have 18PRs—18_250 = 4500 now minus 7 regressions ($125 each), which is 7_125 = 875, so the total is 4500-875=$3625.
> 
> The payment total for [@ZhenjaHorbach](https://github.com/ZhenjaHorbach) for this entire initiative would be $3625. This seems reasonable, given that the project took 8-9 months to complete.
> 
> [@NicMendonca](https://github.com/NicMendonca), could you help us with the payout, please?

Sounds good !
Thanks a lot !

**Comment by NicMendonca** [comment]
### Payment summary

Contributor+: @ZhenjaHorbach - $3,625 paid via NewDot

Payment math laid out [here.](https://github.com/Expensify/App/issues/49354#issuecomment-2989194742)
